### ISSUE TITLE
Gulp slow execution

### ISSUE THREAD

**Issue by grigy** [yeoman/generator-webapp#407]
I'm trying this on Mac with 4GB RAM. It takes ~30 s to start and here is the run log. Is this normal?

[12:06:00] Requiring external module babel-core/register
[12:06:28] Using gulpfile gulpfile.babel.js
[12:06:28] Starting 'clean'...
[12:06:29] Finished 'clean' after 699 ms
[12:06:29] Starting 'default'...
[12:06:29] Starting 'lint'...
[12:06:39] Starting 'styles'...
[12:06:44] Starting 'images'...
[12:06:59] Starting 'fonts'...
[12:07:02] Starting 'extras'...
[12:07:02] Finished 'default' after 33 s
[12:07:02] Finished 'images' after 18 s
[12:07:02] Finished 'lint' after 33 s
[12:07:02] Finished 'styles' after 24 s
[12:07:02] Starting 'html'...
[12:07:26] Finished 'extras' after 25 s
[12:07:27] Finished 'fonts' after 28 s
[12:07:27] Finished 'html' after 25 s
[12:07:27] Starting 'build'...
[12:07:30] build all files 184.19 kB (gzipped)
[12:07:31] Finished 'build' after 3.3 s

**Comment by oskarrough** [comment]
Since the gulpfile > gulpfile.babel.js conversion everything is a lot slower on all the machines I've tested. I tested similar gulpfiles where `gulpfile.js` would take 2 seconds to finish while `gulpfile.babel.js` would take 10 seconds before it even starts.

That said, I couldn't find a way to improve it…

**Comment by callumlocke** [comment]
I often have this problem too, but it's not specific to this generator for me. It's any Node process that reads a lot of files on startup (i.e. most Node processes). And it's very intermittent, sometimes it's a minute, sometimes it's 2 seconds. It seems like sometimes my filesystem is busy and takes ages to respond to anything (maybe there are loads of EMFILEs that are hidden by graceful-fs's quiet retries).

But I don't find it noticeably worse when Babel as a require hook. Can you verify that's your problem?

Try pasting your whole `gulpfile.babel.js` into [Babel](https://babeljs.io/repl/) and then save the output as `gulpfile.js` (and delete the `.babel.js` one to be sure), then run `gulp` again... is there a big speedup?

**Comment by silvenon** [comment]
Do you have a lot of images? Because the `images` task can take a while the first time, before it caches. Does this happen with a fresh install? It takes when I run it ~2.5s.

**Comment by oskarrough** [comment]
I don't believe it has anything to do with images. For instance, the following task takes 3-4 seconds to finish when inside a gulpfile.babel.js on a i7 MacBook Air with SSD, while the gulpfile.js version is instant.

```js
const gulp = require('gulp');
const del = require('del');

gulp.task('clean', cb => {
	del(['.tmp', 'dist']);
	cb();
});
```

Of course, this might not be the most telling test but there's definitely an expected ramp-up time when using babel. Perhaps all of this is fixed in gulp 4?

**Comment by silvenon** [comment]
> Perhaps all of this is fixed in gulp 4?

I'm using gulp 3 and it's fast for me, so it's gotta be something else. Which version of gulp and Node are you using exactly?

**Comment by grigy** [comment]
@silvenon yes, I tried a few times and it's the same. I have only one image file. Versions - gulp: 3.9.0, node: 0.12.7. I would think it's local problem with my machine but I it can't be that slow. Other gulpfiles and brunch run incomparable faster.

**Comment by silvenon** [comment]
I tried on that version of node, but it's still ~700ms. @sindresorhus any ideas?

**Comment by sindresorhus** [comment]
You could use [time-require](https://github.com/jaguard/time-require) to check if any of the requires are making it slow.

**Comment by sindresorhus** [comment]
Babel caches the output between runs, so might be that the cache isn't working somehow, which would make it a lot slower.

**Comment by grigy** [comment]
@sindresorhus thanks, I'll try to check with time-require. Meanwhile I created my own gulpfile, very minimal with only what I need and without babel. It runs under 1s.

**Comment by monzonj** [comment]
It might be something related to babel. In my case it's so slow I'm planning to go back to ES5.

Notice how it takes a lot of time to just get started:

$ gulp
[17:40:**24**] Requiring external module babel-core/register
[17:40:**38**] Using gulpfile gulpfile.babel.js

**Comment by li-cn** [comment]
+1 encountering this issue too. I've been experiencing inconsistent delay times as well, anywhere from 5s to 2 minutes. Using gulp 3.9.0 and node 0.12.7.

**Comment by silvenon** [comment]
Could someone try if updating babel-core to v6 helps?

```
$ npm install --save-dev babel-core@latest babel-register babel-preset-es2015
```

Then add `.babelrc`:

```
{
  "presets": [
    "es2015"
  ]
}
```

Now try to run a gulp command.

**Comment by monzonj** [comment]
No difference with Babel 6. It feels even slower to be honest

**Comment by neilhem** [comment]
@monzonj If you use node version 4+ try to rename gulpfile.babel.js -> gulpfile.js & change these lines of code

``` js
'use strict'
const gulp = require('gulp');
const gulpLoadPlugins = require('gulp-load-plugins');
const browserSync = require('browser-sync');
const del = require('del');
const stream = require('wiredep').stream;
```

**Comment by jahvi** [comment]
Can confirm it still happens even when running just the lint task it takes 4 times as much compared to the non ES2015/Babel gulpfile (2000ms vs 500ms in my case). It might not be related to the generator maybe just the extra overhead that's required to transpile ES2015?

**Comment by pwang2** [comment]
Thank @neilhem , it works well.  I think it does not worth it to taste babel sweetness here.

**Comment by dmitrylabs** [comment]
Another solution:
`npm install -g babel-cli` (if not installed)
then
`babel gulpfile.babel.js -o gulpfile.js`

**Comment by silvenon** [comment]
In the PR above I implemented @neilhem's solution. Thanks for the suggestions everyone, that should take care of this problem.

**Comment by silvenon** [comment]
Version bump is coming soon.

**Comment by silvenon** [comment]
Released in [v2.1.0](https://github.com/yeoman/generator-webapp/releases/tag/v2.1.0).
